node/src/transaction_pool/candidate/transaction_pool_candidate_state.rs

```diff
@@ -203,10 +209,59 @@ impl TransactionPoolCandidatesState {
         self.by_peer.entry(peer_id).or_default().insert(hash, state);
     }
 
-    pub fn get_batch_to_verify(&self) -> Option<(PeerId, Vec<TransactionWithHash>)> {
-        for hash in self.by_hash.keys() {
+    pub fn transactions_received(
+        &mut self,
+        time: Timestamp,
+        peer_id: PeerId,
+        transactions: Vec<TransactionWithHash>,
+        message_id: P2pNetworkPubsubMessageCacheId,
+    ) {
+        let transaction_hashes = transactions
+            .iter()
+            .map(TransactionWithHash::hash)
+            .cloned()
+            .collect::<Vec<_>>();
+
+        self.by_message_id
+            .insert(message_id, (peer_id, transaction_hashes));
+
+        transactions.into_iter().for_each(|transaction| {
+            self.transaction_received(time, peer_id, transaction);
+        })
+    }
+
+    pub fn get_batch_to_verify(
```

Comment:
This needs more comments, or for the inner blocks to be better abstracted into their own methods so that it is very clear what is going on. Right now the logic is quite complicated and not obvious.